spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkParquetReaders.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.vector.NullCheckingForGet;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.arrow.ArrowAllocation;
+import org.apache.iceberg.arrow.vectorized.VectorizedArrowReader;
+import org.apache.iceberg.parquet.TypeWithSchemaVisitor;
+import org.apache.iceberg.parquet.VectorizedReader;
+import org.apache.iceberg.types.Types;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Type;
+
+public class VectorizedSparkParquetReaders {
+
+  private VectorizedSparkParquetReaders() {
+  }
+
+  public static ColumnarBatchReader buildReader(
+      Schema expectedSchema,
+      MessageType fileSchema,
+      Integer recordsPerBatch) {
+    return (ColumnarBatchReader)
+        TypeWithSchemaVisitor.visit(expectedSchema.asStruct(), fileSchema,
+            new VectorizedReaderBuilder(expectedSchema, fileSchema, recordsPerBatch));
+  }
+
+  private static class VectorizedReaderBuilder extends TypeWithSchemaVisitor<VectorizedReader<?>> {
+    private final MessageType parquetSchema;
+    private final Schema icebergSchema;
+    private final BufferAllocator rootAllocator;
+    private final int batchSize;
+
+    VectorizedReaderBuilder(
+        Schema expectedSchema,
+        MessageType parquetSchema,
+        int bSize) {
+      this.parquetSchema = parquetSchema;
+      this.icebergSchema = expectedSchema;
+      this.batchSize = bSize;
+      this.rootAllocator = ArrowAllocation.rootAllocator()
+          .newChildAllocator("VectorizedReadBuilder", 0, Long.MAX_VALUE);
+    }
+
+    @Override
+    public VectorizedReader<?> message(
+            Types.StructType expected, MessageType message,
+            List<VectorizedReader<?>> fieldReaders) {
+      return struct(expected, message.asGroupType(), fieldReaders);
+    }
+
+    @Override
+    public VectorizedReader<?> struct(
+            Types.StructType expected, GroupType struct,
+            List<VectorizedReader<?>> fieldReaders) {
+
+      Map<Integer, VectorizedReader<?>> readersById = Maps.newHashMap();
+      List<Type> fields = struct.getFields();
+
+      IntStream.range(0, fields.size())
+          .forEach(pos -> readersById.put(fields.get(pos).getId().intValue(), fieldReaders.get(pos)));
+
+      List<Types.NestedField> icebergFields = expected != null ?
+          expected.fields() : ImmutableList.of();
+
+      List<VectorizedReader<?>> reorderedFields = Lists.newArrayListWithExpectedSize(
+          icebergFields.size());
+
+      for (Types.NestedField field : icebergFields) {
+        int id = field.fieldId();
+        VectorizedReader<?> reader = readersById.get(id);
+        if (reader != null) {
+          reorderedFields.add(reader);
+        } else {
+          reorderedFields.add(VectorizedArrowReader.nulls());
```

Comment:
Coverage shows that this branch is never taken because there are no tests for read projection that use the vectorized path. Can you add tests based on `TestReadProjection` like the tests you added using `AvroDataTest`?

Reply:
Coverage shows that this is now tested with the update in my PR.